lib/Support/Progress.cpp

```diff
@@ -0,0 +1,255 @@
+
+public:
+  void handleNewTask(const llvm::Task *T) override {
+    LastStepName.emplace_back();
+    emitEvent(T->name(), "task", "B");
+  }
+
+  void handleTaskCompleted(const llvm::Task *T) override {
+    endStep();
+    LastStepName.pop_back();
+    emitEvent(T->name(), "task", "E");
+  }
+
+  void handleTaskAdvancement(const llvm::Task *T,
+                             llvm::StringRef NewStepName) override {
+    endStep();
+    emitEvent(NewStepName, "task", "B");
+    LastStepName.back() = NewStepName;
+  }
+
+private:
+  void endStep() {
+    if (LastStepName.back().has_value()) {
+      emitEvent(*LastStepName.back(), "task", "E");
+      LastStepName.back() = std::nullopt;
+    }
+  }
+
+  void emitEvent(llvm::StringRef Name,
+                 llvm::StringRef Category,
+                 llvm::StringRef Phase) {
+    Output << "{";
+    Output << "\"name\": \"" << Name.str() << "\", ";
+    Output << "\"cat\": \"" << Category.str() << "\", ";
+    Output << "\"ph\": \"" << Phase.str() << "\", ";
+    using namespace std::chrono;
+    unsigned long long Timestamp = duration_cast<std::chrono::microseconds>(system_clock::now().time_since_epoch()).count();
+    Output << "\"ts\": " << Timestamp << ", ";
+    Output << "\"pid\": " << getpid() << ", ";
+    Output << "\"tid\": " << getpid();
+    Output << "},\n";
+  }
+
+};
+
+class PlainProgressListener : public llvm::ProgressListener {
+private:
+  llvm::raw_ostream &Output;
+
+public:
+  PlainProgressListener(llvm::raw_ostream &Output) : Output(Output) {}
+
+public:
+  void handleNewTask(const llvm::Task *T) override {
+    indent(T->index());
+    Output << "Starting " << T->name().str();
+    auto MaybeStepsCount = T->totalSteps();
+    if (MaybeStepsCount)
+      Output << " (" << *MaybeStepsCount << ")";
+    Output << "\n";
+  }
+
+  void handleTaskCompleted(const llvm::Task *T) override {
+    indent(T->index() + 1 - 1);
+    Output << "Ending " << T->name().str() << "\n";
+  }
+
+  void handleTaskAdvancement(const llvm::Task *T,
+                             llvm::StringRef NewStepName) override {
+    indent(T->index() + 1);
+    Output << NewStepName.str() << "\n";
+  }
+
+private:
+  void indent(unsigned Size) {
+    for (unsigned I = 0; I < Size; ++I)
+      Output << "  ";
+  }
+
+};
+
+inline std::string pad(const llvm::Twine &String, const size_t Size, char PaddingChar=' ') {
+  std::string Result = String.str();
+
+  if(Size > Result.size())
+    Result.insert(0, Size - Result.size(), PaddingChar);
+
+  return Result;
+}
+
+
+class TerminalBarsProgressListener : public llvm::ProgressListener {
+private:
+  size_t MaxProgressBars = 0;
+  llvm::raw_ostream &Output;
+  std::chrono::time_point<std::chrono::high_resolution_clock> LastDraw;
+
+public:
+  TerminalBarsProgressListener(llvm::raw_ostream &Output)
+    : Output(Output) {}
+
+public:
+  void handleNewTask(const llvm::Task *T) override {
+    if (T->stack().Tasks.size() > MaxProgressBars) {
+      ++MaxProgressBars;
```

Comment:
:thinking: `MaxProgressBars` is not safe to be accessed concurrently. What happens if 2 threads call `handleNewTask` on the same listener concurrently?

Reply:
Please do not review this, this is draft and WIP, I still need to work out multi-threading details of these guys.